### PR TITLE
Fix expected trip repetition bug, add test to prevent it happening again

### PR DESCRIPTION
**Pull request by GabrielKS**
Fixes https://github.com/e-mission/e-mission-docs/issues/654

### CODE REVIEW

File: emission/tests/pipelineTests/TestPipelineStageNonrepetition.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        eaue._test_options = test_options_stash
+
+    def reset_all(self):
+        etc.dropAllCollections(edb._get_current_db())
+    
+    def count_keys(self):
+        counts = {}
+        db = edb.get_analysis_timeseries_db()
+        for key in self.keys_to_track:
+            count = db.find({"metadata.key": key, "user_id": self.testUUID}).count()
```
Comment (shankari) [inline emission/tests/pipelineTests/TestPipelineStageNonrepetition.py:66]: I believe that `count()` is deprecated and replaced by `countDocuments()`

Comment (shankari) [inline emission/tests/pipelineTests/TestPipelineStageNonrepetition.py:66]:
```suggestion
            count = db.countDocuments({"metadata.key": key, "user_id": self.testUUID})
```

**Review by shankari — CHANGES_REQUESTED** [review]
Can you add some information on what the problem as and how this fixes the it? From the change, it looks like we would have got `expected_trip = None`. So when we tried to access `expected_trip["data"]["end_ts"]`, we would have got an exception.

But that code is outside the `try/expect` block so it would not be handled. So we should have logged that exception and generated an error. But I didn't see any errors in the intake stdinout. What am I missing here?

**Comment by GabrielKS** [comment]
> Can you add some information on what the problem as and how this fixes the it? From the change, it looks like we would have got `expected_trip = None`. So when we tried to access `expected_trip["data"]["end_ts"]`, we would have got an exception.
> 
> But that code is outside the `try/expect` block so it would not be handled. So we should have logged that exception and generated an error. But I didn't see any errors in the intake stdinout. What am I missing here?

The bug was that `_process_and_save_trip` did not return a trip. This meant that `expected_trip` from line 23 was always `None` and thus `last_trip_done` from line 17 was always `None`. So the check on line 24 always lazily evaluated to `True` and we never queried `expected_trip["data"]["end_ts"]` to get that error. When we passed `None` as a trip to `mark_expectation_population_done`, it internally handled that as a special case (per the boilerplate I copypasted from the other `mark_..._done`s), so we didn't get an error there either.

File: emission/tests/pipelineTests/TestPipelineStageNonrepetition.py (hunk 1 above)
Comment (GabrielKS) [reply emission/tests/pipelineTests/TestPipelineStageNonrepetition.py]: You're right (though it's `count_documents()`). Fixed now.

**Review by shankari — APPROVED** [review]

**Comment by shankari** [comment]
Restating https://github.com/e-mission/e-mission-server/pull/834#issuecomment-902341047 in my own words to help me recollect it later:

Without the fix, the code is essentially:

```
>>> last_trip_done = None
>>> expected_trip = None
>>> if last_trip_done is None or last_trip_done["data"]["end_ts"] < expected_trip["data"]["end_ts"]:
...     print(True)
True
```

As we can see, it does not error out due to the short circuit. Then, instead of the `print(True)`, `last_trip_done` remains None. So this code gets triggered:

```
        if last_trip_done is None:
            logging.debug("After run, last_trip_done == None, must be early return")
        epq.mark_expectation_population_done(user_id, last_trip_done)
```

**Pull request merged**